### ISSUE TITLE
CloudKit discussion

### ISSUE THREAD

**Issue by relatedcode** [relatedcode/ParseAlternatives#35]

**Comment by Burnsoft** [comment]
so far cloudkit is looking very good for my needs as a parse alternative.

Only downside I;m working through is understanding security roles and how to set each use with default read/write permissions to the public database.

as I need all users to be able to create records and have others edit them.

Currently I can only see this can be done on a per user basis in the dashboard. Which isn't scalable.

anyone had this problem? is there a radar? cheers

**Comment by pardel** [comment]
Could you elaborate on "have others edit them"? What is the definition of others? All users?

**Comment by pardel** [comment]
The dashboard has a public records section and a private one (which is user specific).

**Comment by Burnsoft** [comment]
in parse, any user can edit an existing object.
In cloudkit, default permissions means only the creator of a record can edit that record.
You can manually add users to a security group to enable them to be able to edit all records of that type, but I wish for the default setting to be that any user can edit the record created by any other user.

I'm creating records that once exist, only certain fields within that record need to be amended. 
The only workaround aside from editing each users ACL in the dashboard, is to create a new record each team a new user creates this record. which is just wasteful.
hope that makes sense.

**Comment by Burnsoft** [comment]
<img width="285" alt="screen shot 2016-01-31 at 14 38 01" src="https://cloud.githubusercontent.com/assets/189002/12702715/55f1c372-c828-11e5-84d9-b70b4d541106.png">

this could be the best/only? way to do this I guess.

On the Record Types, amend the security setting of the record you wish everyone to be able to edit as the above settings.
This way authenticated users can create and edit their own and other records by default.

**Comment by pardel** [comment]
I would separate the read-only part of the model from the fields that will be changing and reference the 2 resulting records.
Also, you need to be careful with the atomicity of the saves - there is no guarantee on the public database.

You can give everyone access to those records and design your app to enforce the roles. And, as you'll have to use the public database, you'll have also to factor in the costs as you're going to pay them.

**Comment by Burnsoft** [comment]
thanks. you mean costs as in the payment to apple for using the public database? database size will be low for me, so I don't expect to incur any costs.

you mention a way to design my app to enforce roles? Do you have any example of this, I've not come across this in the docs, many thanks

**Comment by pardel** [comment]
Yes. You as a dev are expected to cover the public db costs and each user, her/his private db.

Re roles: You don't use roles per se but structure your business logic to allow edits for authenticated users. Also, each record has the creator's ID associated with it so easy to determine ownership:
<img width="882" alt="screen shot 2016-02-02 at 04 41 20" src="https://cloud.githubusercontent.com/assets/2171/12740040/54166f68-c967-11e5-8623-62feee5bf8d3.png">
)

**Comment by Burnsoft** [comment]
even if I get to 100k users, I won't be transferring that much data. when Apple say its free with very high limits, they really do mean high limits :-)
<img width="818" alt="screen shot 2016-02-02 at 13 36 36" src="https://cloud.githubusercontent.com/assets/189002/12750921/031937a8-c9b2-11e5-8986-9dbc32efdfe7.png">

**Comment by fvimagination** [comment]
What about Users? We need to build a custom Record Type called **User_** for example, in order to build a Login/Sign Up controller like we were easily doing with Parse PFUser.
We can't freely access or delete User Records, as far as we could see in CloudKit...

**Comment by Burnsoft** [comment]
you should be able to http://stackoverflow.com/a/35162859/235647

**Comment by jeffreyjackson** [comment]
Of course this is limited to only Apple OSes, whereas Parse provides SDKs for multiple types (.NET, Android, Raspberry Pi, Arduino)

**Comment by agibson73** [comment]
Just throwing my opinion in. CloudKit is super capable but or predicates are not supported so I had to use arrays for storing objects that I would query with or queries in Parse. Also the User cannot be queried as far as I can tell so I am using my own class for this. Hopefully I am wrong and someone can tell me what is wrong. Finally CKAssets can be very slow. I wish they would expose a public url to use SDWebImage. Right now I am saving to an AWS url and my speeds are as fast as Parse. I have had to change my way of thinking but it is working well now.

**Comment by fvimagination** [comment]
Yes, agibson73, we've experienced the same issues, and had to create a custom Users Record Type as well. Anyway, may you post an example of your alternative method to make OR queries with CloudKit?
Thanks!

**Comment by agibson73** [comment]
Fvimagination - Well in my case I was dealing with a game object at parse that had gameuser1 and 2. I was doing a query for all games where the current user was either game user 1 or 2. In CloudKit this did not work as a compound query with other parameters so I am storing both users in as ckreference in nsarry on the object and now I am using [user] in predicate to get all records that the user is in the game. Another interesting thing is that Parse would even give you ref objects and CloudKit makes you do another call. So I am converting these objects to dictionaries in string formats amd storing on the object and parsing them to limit calls. Crazy work around a though.  One finally thing is I have created a local user class that blocks everything until I can check defaults for a logged in user and set that user up. That way I can check for myUserClass.currentUser != nil on setups and other calls just like parse. If not send to Login. Thoughts on what others are doing would be great as well. Notifications are also interesting as the message has to be constructed client side. Fun stuff right.

**Comment by Burnsoft** [comment]
http://stackoverflow.com/a/30952797/235647

bugger. I couldn't find a radar on this, so I've created this one
24564251- OR predicates not accepted in CloudKit

feel free to dupe it.

**Comment by fvimagination** [comment]
Great, thanks do much!

**Comment by fvimagination** [comment]
Last question: anybody has a good tutorial for **Push Notifications** with CloudKit? I've tried some but I haven't understood the "_concept_" behind Subscriptions well, I was able to store Subscriptions in my dashboard for a specific Record Type (example, LIKES class, when a user likes a post of another user, a push notification should be fired to that last user) , but no notifications on my device. APN certificates are both valid and have been downloaded. 
Thanks again :)

**Comment by Burnsoft** [comment]
try this one, I found it really helpful
https://www.invasivecode.com/weblog/advanced-cloudkit-i

**Comment by fvimagination** [comment]
Thanks, I've tried it with no success, probably I did something wrong in the code, I'll check it out again, thanks!

**Comment by agibson73** [comment]
So subscriptions don't have to be stored. I don't know how to explain because I don't know exactly what apple does on the backend. You also don't need APN setup. So that user needs to call a subscription with a predicate of postid == "userPost". Assuming this exists on the Like Object. Set it for records change to fire and subscribe and it will fire ever time the record is changed. All handled by Apple. I will try to post some code when I am back around my computer.

**Comment by fvimagination** [comment]
Ok, sounds great, thanks. The only thing is that we have a custom Record Type called USERS_ (not using the standard **Users**), and we would need to push a notification to a specific user of the class USERS_, not a generic notifications for ALL registered devices, you know what I mean?

**Comment by agibson73** [comment]
yes. The predicate in the subscription determines this. You could even get creative and have the predicate be built by a dynamic record value provided from another CKRecord in changing it you unsubscribe the old subscription when doing a new subscriptions.  I have a string that get sent to the opponent of a game every time the opponent sends a game turn. The string is dynamic in that I have it constructed from a CKRecord so that for holidays and such and localization I can change this. Give me an example and in a bit I will drop an example that might fight your scenario. Like how does the subscription know the users that need it. Is that dynamic or based on a property on the USER_. I myself and using a custom user class so no worries.

**Comment by fvimagination** [comment]
Yes, it should be based on the 'username' field of my **USERS_**. 
For example, my app has a recipe details screen with a like button. That recipe's owner is _sarahdoe_, I'm _johndoe_, so when i tap the like button, the app saves that like into the Likes record type (which has a Reference field for the User.recordName and the Recipe.recordName). If error=nil, i need a piece of code in there to fire a push notification to _sarahdoe_ with a string that of course I'm able to build, like "johndoe liked your recipe: Big Cake".

**Comment by Burnsoft** [comment]
part 3 of this series is all about subscriptions, incase you missed it:
https://www.invasivecode.com/weblog/advanced-cloudkit-part-iii

**Comment by fvimagination** [comment]
Yep, I've tried that too, with no success :(

**Comment by stevenschafer** [comment]
Yea the user management with parse + FBlogin button was so simple. I'm currently also creating a custom USER_ record to model this.
I must say I'm a bit lost on the right way to go about this. I'm new to all this development stuff (Product Designer here) and Parse made logging in with Facebook (and subsequently creating a PFUser) super easy.

I would love to see examples of how you guys are going about this.

Basically, I want to have a "Log In With Facebook" button which copies their FB information and creates a new CKRecord "USER_" to be able to tell if they are already logged in or not when opening the app. Pulling in other relevant FB data like Real Name and Profile photo is nice too.

**Comment by agibson73** [comment]
FVImagination - Here is a snippet that would work on a recipe class.  You will have to change some property names.

`    func subscribeToLikesFor(recipe:CKRecord) {
        
        if MESUser.currentUser.gameUser != nil{
            let currentUser = // retrieve your ckrecord of the current user
            let currentRef = CKReference(record: currentUser, action: .None)
            
            // get a ref to the recipe
            let RecipeRf = CKReference(record: recipe, action: .None)
            let RecipeRfPredicate = NSPredicate(format: "RecipeRef == %@", currentRef)
            let UploadedByCurrentUser = NSPredicate(format: "PostCreatedBy == %@", currentRef)
         
            let compound = NSCompoundPredicate(andPredicateWithSubpredicates: [RecipeRfPredicate,UploadedByCurrentUser])
            /* now we need to create the options fire firing.  This means everytime a like object is changed the user gets a notification*/
            let options: CKSubscriptionOptions = [CKSubscriptionOptions.FiresOnRecordUpdate]
            
            let notificationInfo = CKNotificationInfo()
            // property of record you can change
            // might be an array or just a string
            // "Johnny Liked Your Reciped" Set in PersonLiked and will show to User
            // You don't absolutely have to send anything from the record but this is it
            let properties = ["PersonLiked"]
        
            
            notificationInfo.shouldBadge = true
            //notificationInfo.soundName = UILocalNotificationDefaultSoundName
            notificationInfo.alertLocalizationArgs = properties
            notificationInfo.alertLocalizationKey = "%@"
            notificationInfo.shouldSendContentAvailable = true

            let subscription = CKSubscription(recordType: "Like", predicate: UploadedByCurrentUser, options: options)
            
            subscription.notificationInfo = notificationInfo
            container.publicCloudDatabase.saveSubscription(subscription, completionHandler: {
                subscription, error in
                
                if error == nil{
                    print("subscribed!")
                    
                }
                
            })
        }
    }`

**Comment by fvimagination** [comment]
perfect, thanks, I'll test it out asap!

**Comment by fvimagination** [comment]
Hi again, everybody,
has anyone an idea about how to build a predicate with **OR** clause on CloudKit?
**OR** is not supported by CK, which is pretty bad, so how would you work out a predicate like this?

`    let predicate = NSPredicate(format: "myRecord = %@ OR myRecord = %@", stringA, stringB)
` 

Thanks!

**Comment by agibson73** [comment]
Can you give more context of how you might have an either or recordID?  In
my case I have two users and I decided to store both in an array on the
record so I used CONTAINS.

**Comment by fvimagination** [comment]
the OR predicate I need is to search for roomID into a Messages record type.
when a user sends a message to another user, I store a roomID that's made with the conjunction of userA recordName (the current user) + userB recordName (the receiver), so it's a long string.
in the query I call that string, which can be either userA+userB recordName OR userB+userA recordName.
Hope this helps, and thanks for your reply!

**Comment by agibson73** [comment]
Take my suggestion for what it is worth but I would save the roomID as a
CKReference List(Array).  Then to find that list NSPredicate(format:"ALL %@
in roomID",[CKReference(userA),CKReference(userB)]) Let me know if that
works with your model.

**Comment by fvimagination** [comment]
Sounds great, may you post a sample code about how to save a CKReference List?

**Comment by fvimagination** [comment]
Sounds great, I've worked out a solution based on your comment, and thanks so much again: it works!

**Comment by joevo2** [comment]
Is it possible for this to actually hookup with non apple device?
